### ISSUE TITLE
Automatically overriding a datasource input in loaded dashboards?

### ISSUE THREAD

**Issue by jperville** [PierreVincent/k8s-grafana-watcher#2]
Hello @PierreVincent,

Before using this project, I used to script loading the grafana dashboards like this:

```sh
            for file in *-dashboard.json ; do
              if [ -e "$file" ] ; then
                echo "importing $file" ;
                ( echo '{"dashboard":'; \
                  cat "$file"; \
                  echo ',"overwrite":true,"inputs":[{"name":"DS_PROMETHEUS","type":"datasource","pluginId":"prometheus","value":"prometheus"}]}' ) \
                | jq -c '.' \
                | curl --silent --fail --show-error \
                  --request POST http://${GF_ADMIN_USER}:${GF_ADMIN_PASSWORD}@${GF_SERVICE_NAME}:3000/api/dashboards/import \
                  --header "Content-Type: application/json" \
                  --data-binary "@-" ;
                echo "" ;
              fi
            done
```

How to override automatically the `DS_PROMETHEUS` input using this project? Is there any parameter I can pass to the program to inject the datasource in all of my dashboards?

As of v0.1.0, I can automatically import all my dashboards into Grafana, however the dashboards cannot be used because "Templating init failed Datasource named ${DS_PROMETHEUS} was not found".

**Comment by PierreVincent** [comment]
Hi Julien,

At the moment there isn't, the only thing that can be injected is the `dashboard` field for the POST, with the `overwrite` field set to `true` (see https://github.com/PierreVincent/k8s-grafana-watcher/blob/master/grafanaUpdater.go#L27)

I have seen the same errors recently when exporting a dashboard, which actually exported some input variables (even though I didn't need any), I'm not sure how Grafana decides what's an input for the export/import. Looking at the [Grafana Dashboard API](http://docs.grafana.org/http_api/dashboard/#create-update-dashboard), the `inputs` field doesn't seem to be documented.

Do you have a use for this input or are you essentially using the same ones every time (in which case maybe there is no need to substitute anything)?

I would have to think a bit on how to support the `inputs` field if that was needed however. One option could be to have the full `{ "dashboard": ..., "inputs": ... }` as the body of the ConfigMap key (instead of the content of the `dashboard` field). It would unfortunately not be backward compatible, unless explicitly supported (i.e. check if there is a wrapper object with `dashboard` otherwise assume it's the dashboard definition alone, etc.)

Any thoughts?

**Comment by jperville** [comment]
All my dashboards use the same datasource, ${DS_PROMETHEUS}.

My workflow to produce the dashboards is:
 1. load a dashboard that I like from grafana.net
 2. edit it to my taste
 3. export the customized dashboard as json with the GUI
 4. then I vendor the dashboard in my chef cookbook (git version controlled)
 5. then when I provision a server, chef will load the dashboard content into a k8s configmap
 6. then the grafana-watcher sidecar picksup the dashboard and loads it into grafana

What I like with the current format is that what I put in the grafana-watcher annotated configmap in step 5 is the same json that I exported from grafana in step 3. If we have to wrap the json with `{ "dashboard": ..., "inputs": ... }` then it is not as convenient, even if it can probably be done using chef.

I'd rather have an extra command line argument or environment variable containing json to merge for every dashboard, and keep my dashboard json unchanged.

**Comment by PierreVincent** [comment]
Right I get it now! In your script, your are using is `POST /api/dashboard/import` endpoint, which is actually dedicated to the Dashboard Import function from the UI (see https://github.com/grafana/grafana/issues/7029). The k8s-grafana-watcher is using the `POST /api/dashboards/db`.

Instead of exporting the dashboard from the UI, if you use the "View JSON" and save that instead of the output Export, you should be getting a dashboard definition without the need for the `inputs`.

**Comment by jperville** [comment]
Here is the difference between the same dashboard "view json" and "share dashboard / view json":

```diff
--- view-json.json	2017-08-29 11:05:54.724010571 +0200
+++ export-json.json	2017-08-29 11:06:19.668050109 +0200
@@ -1,4 +1,40 @@
 {
+  "__inputs": [
+    {
+      "name": "DS_PROMETHEUS",
+      "label": "prometheus",
+      "description": "",
+      "type": "datasource",
+      "pluginId": "prometheus",
+      "pluginName": "Prometheus"
+    }
+  ],
+  "__requires": [
+    {
+      "type": "grafana",
+      "id": "grafana",
+      "name": "Grafana",
+      "version": "4.4.1"
+    },
+    {
+      "type": "panel",
+      "id": "graph",
+      "name": "Graph",
+      "version": ""
+    },
+    {
+      "type": "datasource",
+      "id": "prometheus",
+      "name": "Prometheus",
+      "version": "1.0.0"
+    },
+    {
+      "type": "panel",
+      "id": "singlestat",
+      "name": "Singlestat",
+      "version": ""
+    }
+  ],
   "annotations": {
     "list": []
   },
@@ -7,7 +43,7 @@
   "gnetId": 1573,
   "graphTooltip": 0,
   "hideControls": false,
-  "id": 2,
+  "id": null,
   "links": [],
   "refresh": "5m",
   "rows": [
@@ -24,7 +60,7 @@
             "rgba(237, 129, 40, 0.89)",
             "rgba(50, 172, 45, 0.97)"
           ],
-          "datasource": "prometheus",
+          "datasource": "${DS_PROMETHEUS}",
           "format": "none",
           "gauge": {
             "maxValue": 100,
@@ -99,7 +135,7 @@
             "rgba(237, 129, 40, 0.89)",
             "rgba(50, 172, 45, 0.97)"
           ],
-          "datasource": "prometheus",
+          "datasource": "${DS_PROMETHEUS}",
           "format": "none",
           "gauge": {
             "maxValue": 100,
@@ -174,7 +210,7 @@
             "rgba(237, 129, 40, 0.89)",
             "rgba(50, 172, 45, 0.97)"
           ],
-          "datasource": "prometheus",
+          "datasource": "${DS_PROMETHEUS}",
           "decimals": 2,
           "format": "decbytes",
           "gauge": {
@@ -258,7 +294,7 @@
           "bars": false,
           "dashLength": 10,
           "dashes": false,
-          "datasource": "prometheus",
+          "datasource": "${DS_PROMETHEUS}",
           "fill": 1,
           "id": 4,
           "legend": {
@@ -347,7 +383,7 @@
           "bars": false,
           "dashLength": 10,
           "dashes": false,
-          "datasource": "prometheus",
+          "datasource": "${DS_PROMETHEUS}",
           "fill": 1,
           "id": 5,
           "legend": {
@@ -444,7 +480,7 @@
           "bars": false,
           "dashLength": 10,
           "dashes": false,
-          "datasource": "prometheus",
+          "datasource": "${DS_PROMETHEUS}",
           "fill": 1,
           "id": 6,
           "legend": {
@@ -541,7 +577,7 @@
           "bars": false,
           "dashLength": 10,
           "dashes": false,
-          "datasource": "prometheus",
+          "datasource": "${DS_PROMETHEUS}",
           "fill": 1,
           "id": 7,
           "legend": {
@@ -651,11 +687,8 @@
     "list": [
       {
         "allValue": null,
-        "current": {
-          "text": "myhost",
-          "value": "myhost"
-        },
-        "datasource": "prometheus",
+        "current": {},
+        "datasource": "${DS_PROMETHEUS}",
         "hide": 0,
         "includeAll": false,
         "label": "Instance",
```

Notice that:
 - the "id" field is set with "view json" and null with "export json"
 - the current value for a template variable ("myhost") shows up in the "view json", which is not what I want (I would have to manually clean dashboards before I can put them in the chef cookbook).

**Comment by PierreVincent** [comment]
Yep, the `id` would need to be set to `null` after the "View JSON".  The template variable default value doesn't tend to be a problem from what I've experienced, if you use the "refresh on dashboard load" for query template variables... still not very pretty I have to agree there.

I think deep down the issue here is that Export is meant to map with the Import API, as opposed to the create/update API. And the Import API is not documented because solely meant to be used through the UI (from what the Grafana authors say), mostly from the fact that the Import API requires these inputs that are being set at import time.

I don't think there is an easy way around it with the current way the watcher posts the dashboards. Maybe adding another type of annotation that would consider the ConfigMap not as a dashboard definition, but as a templated dashboard definition... they can then be posted to the import endpoint, but it will need get the inputs from somewhere (other ConfigMap? command line arguments? what if the inputs are different for different dashboards?)

I would also be a bit wary of the undocumented nature of `/api/dashboard/import` if there is no guarantee of it being there going forward (especially if it's being considered as a "UI only" endpoint).

Sorry I can't be much more help on that - this watcher is pretty basic as it started as a side-project. We're using it internally but quite clearly there are some rough edges! :(

**Comment by jperville** [comment]
I could live with another annotation to tell "this is a templated dashboard"; I would generate the templated dashboard from the exported dashboard and add the datasource input with chef. Would that be difficult for you to implement?

**Comment by PierreVincent** [comment]
I think it should be pretty straightforward. I can't promise anything but I'll try to take a look at it sometime this week.

**Comment by PierreVincent** [comment]
Done and released in `v0.2.0`
